includes/session_details.js

```diff
@@ -0,0 +1,607 @@
+    TIMESTAMP_DIFF(LEAD(page_visit_at) OVER session_window, page_visit_at, SECOND) AS page_duration_seconds,
+    -- page_duration_seconds is set to the different between the page_entry_time of the current page and the page_entry_time of the following page in the session
+    next_step
+  FROM
+    sessions_grouped
+  WINDOW session_window AS (PARTITION BY session_id ORDER BY page_visit_at)),
+  -- Calculate session level metrics
+  session_metrics AS (
+  SELECT
+    session_id,
+    user_id,
+    -- A session can only have one user_id
+    ARRAY_AGG(STRUCT( anonymised_user_agent_and_ip,
+        -- A session can have multiple anonymised_user_agent_and_ip so this is included in the page level fields.
+        page_path AS page,
+        request_referer_domain as previous_page_domain,
+        previous_page_path AS previous_page,
+        page_entry_time,
+        page_exit_time,
+        page_duration_seconds AS duration,
+        next_step)
+    ORDER BY
+      page_entry_time) AS pages_visited_details,
+    MIN(page_entry_time) AS session_start_timestamp,
+    -- session_start_timestamp is the page_entry_time of the first page
+    MAX(page_entry_time) AS final_session_page_timestamp,
+    -- session_end_timestamp is the page_entry_time of the final page. This is because we cannot know when they left that page as there is no page visit afterwards.
+    COUNT(*) AS count_pages_visited,
+    -- Count of all rows in a session is the pages visited in that session.
+  FROM
+    page_times
+  GROUP BY
+    session_id,
+    user_id)
+  SELECT
+    session_id,
+    user_id,
+    pages_visited_details[0].previous_page_domain AS session_referer_domain,
+    pages_visited_details[0].page AS start_page,
+    -- set start_page to the first page in the pages_visited_details array for a single session
+   ARRAY_REVERSE(pages_visited_details)[0].page AS exit_page,
+    -- set exit_page to the last page in the pages_visited_details array for a single session
+    session_start_timestamp,
+    final_session_page_timestamp,
+    CASE /* This is for single page sessions. By default these will have a session time of 0 but this is not a legitimate value because the start and end time are taken from the same page. Therefore, we set them to NULL so they do not impact calculations of averages */
+      WHEN final_session_page_timestamp = session_start_timestamp THEN NULL
+    ELSE
+    TIMESTAMP_DIFF(final_session_page_timestamp, session_start_timestamp, SECOND)
+  END
+    AS session_time_in_seconds,
+    count_pages_visited,
+    pages_visited_details
+  FROM
+    session_metrics
+  `).preOps((ctx) => {
+    // This pre operation is used to filter the events table to only events that started AFTER the maximum event start time in the current session_details table. 
+    const isIncremental = ctx.incremental(); // Check if the run is incremental
+    const eventTimestampCheckpoint = isIncremental
```

Comment:
This logic could just go inside the ${} on line 603 for brevity I think:
```
DECLARE event_timestamp_checkpoint TIMESTAMP DEFAULT (
      ${ctx.incremental() ? `SELECT MAX(session_start_timestamp) FROM ${ctx.self()}` : `SELECT TIMESTAMP("2000-01-01")`}
    );
```
Then you could miss out the ```const```s

Reply:
Much cleaner - thanks